src/lib.rs

```diff
@@ -99,23 +100,17 @@ impl<T: Eq + Hash + Send + Sync + 'static> ArcIntern<T> {
     /// previously allocated.
     ///
     /// Note that `ArcIntern::new` is a bit slow, since it needs to check
-    /// a `HashMap` protected by a `Mutex`.
+    /// a `DashMap` which contains its own mutexes.
     pub fn new(val: T) -> ArcIntern<T> {
-        let mymutex = Self::get_mutex();
-        let mut m = mymutex.lock().unwrap();
-        if let Some(b) = m.get(&val) {
-            return ArcIntern { arc: b.clone() };
-        }
-        let b = Arc::new(val);
-        let p = ArcIntern { arc: b.clone() };
-        m.insert(b);
-        p
+        let m = Self::get_state();
+        let b = m.entry(Arc::new(val)).or_insert(());
```

Comment:
I couldn't see a way to avoid the Arc creation without introducing a race, as the dashmap entry api needs the key to acquire the shard's mutex.

Looking at the dashmap 4.0 PR https://github.com/xacrimon/dashmap/pull/81  there are quite a few features remaining to be implemented, of which one insert_if_not_exists_and_get sounds like it would be the equivalent operation.  I think its likely it would return a guard similar to https://github.com/xacrimon/dashmap/blob/acrimon/experimental/src/lib.rs#L136.

Reply:
> This is beautiful, thank you so much for this contribution! Do you know, by any chance, how this PR affects the single-threaded performance?

You're welcome.  I think we probably hit the internment race the same week, thanks for the alternative crate!   

I've not run any single threaded benchmarks.  Going by the DashMap benchmark graphs it's probably ok: 
 - [dashmap insert](https://raw.githubusercontent.com/xacrimon/dashmap/master/assets/bench-insert.svg?sanitize=true) 
 - [dashmap get](https://raw.githubusercontent.com/xacrimon/dashmap/master/assets/bench-get.svg?sanitize=true)

Reply:
Thanks for the clarifications! I plan to merge and publish this later today or tomorrow. I wanted to add a sequential performance test first.

PS. I think there is a way to further reduce contention by not taking a lock on each reference decrement, but I haven't had a chance to think this through yet.